crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs

```diff
@@ -721,27 +752,33 @@ impl SasState<WeAccepted> {
     ) -> Result<SasState<KeyReceived>, SasState<Cancelled>> {
         self.check_event(sender, content.flow_id()).map_err(|c| self.clone().cancel(true, c))?;
 
-        let their_pubkey = content.public_key().to_owned();
+        let their_public_key = Curve25519PublicKey::from_slice(content.public_key().as_bytes())
+            .map_err(|_| self.clone().cancel(true, "Invalid public key".into()))?;
 
-        self.inner
-            .lock()
-            .unwrap()
-            .set_their_public_key(their_pubkey.encode())
-            .expect("Can't set public key");
+        let sas = self.inner.lock().unwrap().take();
 
-        Ok(SasState {
-            inner: self.inner,
-            ids: self.ids,
-            verification_flow_id: self.verification_flow_id,
-            creation_time: self.creation_time,
-            last_event_time: Instant::now().into(),
-            started_from_request: self.started_from_request,
-            state: Arc::new(KeyReceived {
-                we_started: self.state.we_started,
-                their_pubkey,
-                accepted_protocols: self.state.accepted_protocols.clone(),
-            }),
-        })
+        if let Some(sas) = sas {
+            let established = sas
+                .diffie_hellman(their_public_key)
+                .map_err(|_| self.clone().cancel(true, "Invalid public key".into()))?;
+
+            Ok(SasState {
+                inner: self.inner,
+                our_public_key: self.our_public_key,
+                ids: self.ids,
+                verification_flow_id: self.verification_flow_id,
+                creation_time: self.creation_time,
+                last_event_time: Instant::now().into(),
+                started_from_request: self.started_from_request,
+                state: Arc::new(KeyReceived {
+                    sas: Mutex::new(established).into(),
+                    we_started: self.state.we_started,
+                    accepted_protocols: self.state.accepted_protocols.clone(),
+                }),
+            })
+        } else {
+            Err(self.cancel(true, CancelCode::UnexpectedMessage))
```

Comment:
Is this right? This seems unrelated to the given message (i.e. `content`) and depends solely on the `self.inner.lock().unwrap().take()` call failing.

Perhaps the result of that call depends on an earlier message? But in any case, it's a bit unobvious what's happening.

Reply:
We transition here from a Sas object into an `EstablishedSas` object using the `diffie_hellman()` method. This  method takes `self` since our private curve25519 key is ephemeral and can only be used once.

This in turn means that we can only transition into the `KeyReceived` state once, if we receive another `m.key.verification.key` message we can't proceed. We can either cancel here or ignore the message.

Reply:
Right, so the fact that `self.inner` is `None` would ultimately indicate that we've received `m.key.verification.key` twice, so the error code is indeed justified. Thanks.